xds/src/main/java/io/grpc/xds/CdsLoadBalancer2.java

```diff
@@ -0,0 +1,328 @@
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            if (shutdown) {
+              return;
+            }
+            // All watchers should receive the same error, so we only propagate it once.
+            if (ClusterState.this == root) {
+              handleClusterDiscoveryError(error);
+            }
+          }
+        });
+      }
+
+      @Override
+      public void onResourceDoesNotExist(String resourceName) {
+        syncContext.execute(new Runnable() {
+          @Override
+          public void run() {
+            if (shutdown) {
+              return;
+            }
+            discovered = true;
+            result = null;
+            if (childClusterStates != null) {
+              for (ClusterState state : childClusterStates.values()) {
+                state.shutdown();
+              }
+              childClusterStates = null;
+            }
+            handleClusterDiscovered();
+          }
+        });
+      }
+
+      @Override
+      public void onChanged(final CdsUpdate update) {
+        class ClusterDiscovered implements Runnable {
+          @Override
+          public void run() {
+            if (shutdown) {
+              return;
+            }
+            discovered = true;
+            result = update.clusterConfig;
+            if (update.clusterType == ClusterType.AGGREGATE) {
+              isLeaf = false;
+              AggregateClusterConfig clusterConfig = (AggregateClusterConfig) update.clusterConfig;
+              logger.log(XdsLogLevel.INFO, "Aggregate cluster {0}", update.clusterName);
+              logger.log(XdsLogLevel.DEBUG, "Cluster config: {0}", clusterConfig);
+              Map<String, ClusterState> newChildStates = new LinkedHashMap<>();
+              for (String cluster : clusterConfig.prioritizedClusterNames) {
+                if (childClusterStates == null || !childClusterStates.containsKey(cluster)) {
+                  ClusterState childState = new ClusterState(cluster);
+                  childState.start();
+                  newChildStates.put(cluster, childState);
+                } else {
+                  newChildStates.put(cluster, childClusterStates.remove(cluster));
+                }
+              }
+              if (childClusterStates != null) {  // stop subscribing to revoked child clusters
+                for (ClusterState watcher : childClusterStates.values()) {
+                  watcher.shutdown();
+                }
+              }
+              childClusterStates = newChildStates;
+            } else if (update.clusterType == ClusterType.EDS) {
+              isLeaf = true;
+              EdsClusterConfig clusterConfig = (EdsClusterConfig) update.clusterConfig;
+              logger.log(XdsLogLevel.INFO, "EDS cluster {0}, edsServiceName: {1}",
+                  update.clusterName, clusterConfig.edsServiceName);
+              logger.log(XdsLogLevel.DEBUG, "Cluster config: {0}", clusterConfig);
```

Comment:
also revoke child clusters if `if (childClusterStates != null)`, and then set `childClusterStates = null`?

Reply:
I don't think we need to consider such a weird use case: a EDS/Logical DNS cluster becomes an aggregate cluster (or vice versa).